### ISSUE TITLE
The source code of the paper TNASP

### ISSUE THREAD

**Issue by tjdeng** [ShunLu91/shunlu91.github.io#1]
Hi,

I have already finished reading your paper about TNASP. That's great work. I want to reproduce the experiments mentioned on TNASP. So when will you release the code? Thanks.

Best

**Comment by ShunLu91** [comment]
Hello,

Thanks for your attention. 
As the code of TNASP is also related to the subsequent work, which is under review, we plan to release them together later.
In fact, our code is heavily based on this repo: https://github.com/ultmaster/neuralpredictor.pytorch
You can try to re-implement our work by adding the Transformer architecture according to the settings in our paper.
Feel free to ask any question and I will reply as soon as possible.

Sincerely

**Comment by tjdeng** [comment]
Thanks.

**Comment by tjdeng** [comment]
Hi, 
I'm re-implementing your work TNASP followed by implementation details at the supplementary. Unfortunately, the result is bad, and the transformer-based predictor is worse than the MLP-based and GCN-based predictor. So could you give me more information about Encoder and Regressor? Now I have two questions as followed:
1. How to combine e1 and e2? I directly add both together.
2. How to change the shape of the fusion feature e to enter the MLP regressor? I concatenate e's dim-N and dim-M, and convert e into a vector in order to get the output with dim-1 in the final layer.

here is my re-implementation:
```
class TransformerPredictor(nn.Module):
    def __init__(self, operation_dim, position_dim, trans_hidden=80, linear_hidden=96):
        super().__init__()
        self.operation_encoding = nn.Linear(operation_dim, trans_hidden, bias=False)
        self.position_encoding = nn.Linear(position_dim, trans_hidden, bias=False)
        encoder_layer = nn.TransformerEncoderLayer(d_model=trans_hidden, nhead=4)
        self.transformer_encoder = nn.TransformerEncoder(encoder_layer, num_layers=3)
        self.dropout = nn.Dropout(0.1)
        self.fc1 = nn.Linear(position_dim*trans_hidden, linear_hidden, bias=False)
        self.fc2 = nn.Linear(linear_hidden, 1, bias=False)

    def forward(self, inputs):
        numv, adj, out = inputs["num_vertices"], inputs["adjacency"], inputs["operations"]
        opt_out = self.operation_encoding(out)
        adj = adj.float()
        pos_out = self.position_encoding(adj)
        trans_input = opt_out + pos_out
        out = self.transformer_encoder(trans_input)
        out = out.view(out.shape[0], -1)
        out = self.fc1(out)
        out = self.dropout(out)
        out = self.fc2(out).view(-1)
        return out
```
I test all exps on NASBench201, so the operation_dim is 7 (one-hot encoding) and the position_dim is 8. After that, we can get e1 and e2 with the shape (batch_size, 8, 80), and add them together to get trans_input. Finally, the output of the transformer is flattened to get the input of the mlp regressor with the shape (batch_size, 640). 

I'm looking forward to your reply, thanks~

**Comment by ShunLu91** [comment]
Thanks for your efforts. I have changed several lines in your code and tested it, here are the results:
```Python
class TransformerPredictorV1(nn.Module):
    def __init__(self, operation_dim, position_dim, trans_hidden=80, linear_hidden=96):
        super().__init__()
        self.operation_encoding = nn.Linear(operation_dim, trans_hidden, bias=False)
        self.position_encoding = nn.Linear(position_dim, trans_hidden, bias=False)
        encoder_layer = nn.TransformerEncoderLayer(d_model=trans_hidden, nhead=4)
        self.transformer_encoder = nn.TransformerEncoder(encoder_layer, num_layers=3)
        self.dropout = nn.Dropout(0.1)
        self.fc1 = nn.Linear((operation_dim+1)*trans_hidden, linear_hidden, bias=False)
        self.fc2 = nn.Linear(linear_hidden, 1, bias=False)

        self.pos_map = nn.Linear(position_dim, operation_dim+1)

    def forward(self, inputs):
        numv, adj, out = inputs["num_vertices"], inputs["adjacency"], inputs["operations"]
        opt_out = self.operation_encoding(out)
        adj = adj.float()

        # map dim
        pos_out = self.pos_map(adj).transpose(1, 2)
        pos_out = self.position_encoding(pos_out)

        trans_input = opt_out + pos_out
        out = self.transformer_encoder(trans_input)
        out = out.view(out.shape[0], -1)
        out = self.fc1(out)
        out = self.dropout(out)
        out = self.fc2(out).view(-1)
        return out
```
- Use TransformerPredictorV1 on NAS-Bench-201 with 78 data pairs, we can achieve the Kendalltau 0.485927.

 ```Python
class TransformerPredictorV2(nn.Module):
    def __init__(self, operation_dim, position_dim, trans_hidden=80, linear_hidden=96):
        super().__init__()
        self.operation_encoding = nn.Embedding(operation_dim, trans_hidden, padding_idx=None)
        self.position_encoding = nn.Linear(position_dim, trans_hidden, bias=False)
        encoder_layer = nn.TransformerEncoderLayer(d_model=trans_hidden, nhead=4)
        self.transformer_encoder = nn.TransformerEncoder(encoder_layer, num_layers=3)
        self.dropout = nn.Dropout(0.1)
        self.fc1 = nn.Linear((operation_dim+1)*trans_hidden, linear_hidden, bias=False)
        self.fc2 = nn.Linear(linear_hidden, 1, bias=False)

        self.pos_map = nn.Linear(position_dim, operation_dim+1)

    def forward(self, inputs):
        numv, adj, out = inputs["num_vertices"], inputs["adjacency"], inputs["features"]
        opt_out = self.operation_encoding(out)
        adj = adj.float()

        # map dim
        pos_out = self.pos_map(adj).transpose(1, 2)
        pos_out = self.position_encoding(pos_out)

        trans_input = opt_out + pos_out
        out = self.transformer_encoder(trans_input)
        out = out.view(out.shape[0], -1)
        out = self.fc1(out)
        out = self.dropout(out)
        out = self.fc2(out).view(-1)
        return out
```
- Use TransformerPredictorV2 on NAS-Bench-201 with 78 data pairs, we can achieve the Kendalltau  0.516618. 
- TransformerPredictorV2 is more like our implemention except that we use Laplacian positional encoding.
- Our code is on the way and will be public soon.
- Here are my running logs:
- [log_tp_v1_201_78.txt](https://github.com/ShunLu91/shunlu91.github.io/files/7768680/log_tp_v1_201_78.txt)
- [log_tp_v2_201_78.txt](https://github.com/ShunLu91/shunlu91.github.io/files/7768683/log_tp_v2_201_78.txt)

**Comment by tjdeng** [comment]
Thank you very much for your reply. I copied the code TransformerPredictorV1 onto my envs and tested it. There are still appearing the same error, which is that the value of Kendalltau is too low, such as 0.009. I tested the same data on the MLP-based predictor, which has the same training and testing process, and it performs well. I logged one part of predicting results on test data for transformer-based predictor and found that all predicting values are the same.

here are logs:
[tranformer_num_init-200-T-23-Dec-at-12-10-13.log](https://github.com/ShunLu91/shunlu91.github.io/files/7769320/tranformer_num_init-200-T-23-Dec-at-12-10-13.log)
[MLP_num_init-200-T-23-Dec-at-12-01-25.log](https://github.com/ShunLu91/shunlu91.github.io/files/7769339/MLP_num_init-200-T-23-Dec-at-12-01-25.log)

here is the code about training and testing for both MLP-predictor and transformer-predictor.
```python
def fit(net,
        xtrain,
        gcn_hidden=144,
        seed=0,
        batch_size=7,
        epochs=300,
        lr=1e-4,
        wd=3e-4,
        logger=None
        ):
    reset_seed(seed)
    data_loader = DataLoader(xtrain, batch_size=batch_size, shuffle=True, drop_last=True)
    # net.to(device)
    criterion = nn.MSELoss()
    optimizer = optim.Adam(net.parameters(), lr=lr, weight_decay=wd)
    lr_scheduler = optim.lr_scheduler.CosineAnnealingLR(optimizer, epochs)
    net.train()
    logger.log("The form of one xtrain data : {:}".format(xtrain[0]))
    for epoch in range(epochs):
        meters = AverageMeterGroup()
        lr = optimizer.param_groups[0]["lr"]
        for step, batch in enumerate(data_loader):
            target = batch["val_acc"].float()
            prediction = net(batch)
            loss = criterion(prediction, target)
            loss.backward()
            optimizer.step()
            mse = accuracy_mse(prediction, target)
            meters.update({"loss": loss.item(), "mse": mse.item()}, n=target.size(0))
        print("epoch: {}, loss: {}".format(epoch, loss.item()))
        logger.log("epoch: {}, loss: {}".format(epoch, loss.item()))
        lr_scheduler.step()

def get_test_loss(net, xtest, eval_batch_size=1000, logger=None):
    test_data_loader = DataLoader(xtest, batch_size=eval_batch_size)
    criterion = nn.MSELoss()
    net.eval()
    logger.log("The form of one xtest data : {:}".format(xtest[0]))
    predict_, target_ = [], []
    with torch.no_grad():
        for step, batch in enumerate(test_data_loader):
            target = batch["val_acc"].float()
            predict = net(batch)
            predict_.append(predict.numpy())
            target_.append(target.numpy())
            # meters.update({"loss": criterion(predict, target).item(),
            #                "mse": accuracy_mse(predict, target).item()}, n=target.size(0))

            # if (args.eval_print_freq and step % args.eval_print_freq == 0) or \
            #         step % 10 == 0 or step + 1 == len(test_data_loader):
            #     logger.info("Evaluation Step [%d/%d]  %s", step + 1, len(test_data_loader), meters)
    predict_ = np.concatenate(predict_)
    target_ = np.concatenate(target_)
    logger.log("one part of predicting results: {:}".format(predict_[:10]))
    logger.log("target_: {:}".format(target_[:10]))
    result = kendalltau(predict_, target_)[0]
    return result
```

The transformer predictor:
```python
def trans_predictor(
        all_arch_info,
        num_init=200,
        logger=None
    ):

    train_data = random.sample(all_arch_info, num_init)
    xtrain = [d['encoding'] for d in train_data]
    test_data =[v for v in all_arch_info if v not in train_data]
    xtest = [d['encoding'] for d in test_data]

    print("Start to train the predictor ... ")
    # train the neural network
    operation_dim = len(xtrain[0]['operations'][0])
    position_dim = len(xtrain[0]['adjacency'][0])
    net = TransformerPredictor(operation_dim=operation_dim, position_dim=position_dim)
    seed = np.random.choice(10000)
    trans_fit(net, xtrain, seed=seed, logger=logger)
    print("Training successfully !!! ")
    # make predictions on the test data
    test_acc_loss = trans_get_test_loss(net, xtest, logger=logger)
    return test_acc_loss
```

The mlp-predictor:
```python
def mlp_predictor(
        all_arch_info,
        num_init=200,
        logger=None
):

    train_data = random.sample(all_arch_info, num_init)
    xtrain = [d['encoding'] for d in train_data]
    test_data =[v for v in all_arch_info if v not in train_data]
    xtest = [d['encoding'] for d in test_data]

    print("Start to train the predictor ... ")
    # train the neural network
    adj = np.concatenate(xtrain[0]['adjacency'])
    opts = np.concatenate(xtrain[0]['operations'])
    input_dim = len(adj) + len(opts)

    net = MlpPredictor(input_dim=input_dim)
    seed = np.random.choice(10000)
    mlp_fit(net, xtrain, seed=seed, logger=logger)

    print("Training successfully !!! ")
    # make predictions on the test data
    test_acc_loss = mlp_get_test_loss(net, xtest, logger=logger)

    return test_acc_loss
```
Now I realize that the results of predicting are the same value for transformer-predictor. So are you facing the same question? or can you find where is wrong? Thanks.

**Comment by ShunLu91** [comment]
I guess maybe the error lies in the data processing. 
In our code, the target acc is first normalized into [0, 1] and then shifted by ((acc -mean) / std). 
I have met the same error when I used traditional machine learning algorithm to regress the accuracy.
We will release our code as soon as possible.

**Comment by tjdeng** [comment]
Hi, 
I fixed the above bug but the transformer-based predictor performs badly. 
here are the code details of the transformer-based predictor and logs of the gcn-based and transformer-based predictor  on nasbench101:
```python

def get_laplacian_matrix(adj):
    rowsum = adj.sum(2)
    degree_matrix = torch.zeros(adj.shape, device=adj.device)
    for i in range(len(adj)):
        degree_matrix[i] = torch.diag(rowsum[i])
    return torch.sub(degree_matrix, adj)

class TransformerPredictorV2(nn.Module):
    def __init__(self, operation_dim, position_dim, trans_hidden=80, linear_hidden=96):
        super().__init__()
        # self.operation_encoding = nn.Linear(operation_dim, trans_hidden, bias=False)
        self.operation_encoding = nn.Embedding(operation_dim, trans_hidden, padding_idx=None)
        self.position_encoding = nn.Linear(position_dim, trans_hidden, bias=False)
        encoder_layer = nn.TransformerEncoderLayer(d_model=trans_hidden, nhead=4)
        self.transformer_encoder = nn.TransformerEncoder(encoder_layer, num_layers=3)
        self.dropout = nn.Dropout(0.1)
        self.fc1 = nn.Linear((operation_dim+2)*trans_hidden, linear_hidden, bias=False)
        self.fc2 = nn.Linear(linear_hidden, 1, bias=False)

        self.pos_map = nn.Linear(position_dim, operation_dim+2)

    def forward(self, inputs):
        numv, adj, out = inputs["num_vertices"], inputs["adjacency"], inputs["operations"]

        out = out.long()
        adj = get_laplacian_matrix(adj)
        opt_out = self.operation_encoding(out)
        # map dim
        pos_out = self.pos_map(adj).transpose(1, 2)
        pos_out = self.position_encoding(pos_out)

        trans_input = opt_out + pos_out
        out = self.transformer_encoder(trans_input)
        out = out.view(out.shape[0], -1)
        out = self.fc1(out)
        out = self.dropout(out)
        out = self.fc2(out).view(-1)
        return out
```
[gcn_num_init172_2021-12-28 10:21:48.log](https://github.com/ShunLu91/shunlu91.github.io/files/7782512/gcn_num_init172_2021-12-28.10.21.48.log)
[transformer_num_init172_2021-12-28 10:27:00.log](https://github.com/ShunLu91/shunlu91.github.io/files/7782516/transformer_num_init172_2021-12-28.10.27.00.log)

The other coding info is corresponding to the repo that you suggested. Could you give me some advice?

By the way, I  found that it's very important to normalize the val accuracy in exps. here are logs of the gcn-based predictor with normalization and without it respectively:

[gcn_num_init172_2021-12-28 10:21:48.log](https://github.com/ShunLu91/shunlu91.github.io/files/7782521/gcn_num_init172_2021-12-28.10.21.48.log)
[gcn_nonorm_num_init172_2021-12-28 10:53:14.log](https://github.com/ShunLu91/shunlu91.github.io/files/7782523/gcn_nonorm_num_init172_2021-12-28.10.53.14.log)

do you know how to get the mean and std, like the value on nasbench101 in the repo that you suggested? it seemed that the val acc and test acc were normalized with the same mean and std. Besides, how do you set the mean and std on nasbench201?
Thanks.

**Comment by ShunLu91** [comment]
- OK, such results seems to be normal. Actually, I also found that different normalizations led to different results.
- I simply use all the data to calculate the mean and std. 
- Here I provide the dataloader I used as below and you can have a try. Expect your better results.
[release_dataset.txt](https://github.com/ShunLu91/shunlu91.github.io/files/7782621/release_dataset.txt)

**Comment by tjdeng** [comment]
I followed your code of processing data and tested an exp. The result was 0.474197 (Kendalltau) and could not achieve the performance that is shown in your paper table 5. Note that I use my implementation of computing the Laplacian matrix.
I have committed the code to my Github repo. Welcome to check it.

**Comment by ShunLu91** [comment]
Some  details of model definition are still different. I test your code with our model using 'adj' positional encoding and got 0.642167(Kendalltau), which is consistent with our results in Tab.4.
Here is the code and hope your better results.
[Test_neuralpredictor-main.zip](https://github.com/ShunLu91/shunlu91.github.io/files/7784963/Test_neuralpredictor-main.zip)
.
